examples/no_std/rp2040-embassy/src/noline_async.rs

```
@@ -0,0 +1,142 @@
+use core::fmt::Write as FmtWrite;
+use embassy_futures::block_on;
+use embassy_rp::usb::{Driver, Instance};
+use embassy_usb::{
+    class::cdc_acm::{ControlChanged, Receiver, Sender},
+    driver::EndpointError,
+};
+use fixed_queue::VecDeque;
+use noline::{async_io::IO, builder::EditorBuilder};
+
+fn map_error(_e: EndpointError) -> embedded_io_async::ErrorKind {
+    // TODO: This could be more specific
+    embedded_io_async::ErrorKind::Other
+}
+
+// Implement the reader struct
+struct Reader<'d, I: Instance> {
+    stdin: &'d mut Receiver<'d, Driver<'d, I>>,
+    queue: VecDeque<u8, 64>,
+}
+
+// Exposte the wait_connection function from the borrowd stdin
+impl<'d, R: Instance> Writer<'d, R> {
+    fn ready(&self) -> bool {
+        self.stdout.rts() && self.stdout.dtr()
+    }
+    async fn wait_connection(&mut self) {
+        self.stdout.wait_connection().await
+    }
+}
+
+impl<'d, R: Instance> embedded_io_async::ErrorType for Reader<'d, R> {
+    type Error = embedded_io_async::ErrorKind;
+}
+
+// Read data from the input and make it available asynchronously
+impl<'d, R: Instance> embedded_io_async::Read for Reader<'d, R> {
+    async fn read(&mut self, buf: &mut [u8]) -> Result<usize, Self::Error> {
+        // If the queue is empty
+        while self.queue.is_empty() {
+            let mut buf: [u8; 64] = [0; 64];
+            // Read a maximum of 64 bytes from the ouput
+            let len = self
+                .stdin
+                .read_packet(&mut buf)
+                .await
+                .map_err(|e| map_error(e))?;
+            // This is safe because we only ever pull data when empty
+            // And the queue has the same capacity as the input buffer
+            for i in buf.iter().take(len) {
+                let _ = self.queue.push_back(*i);
```

Add `expect()` instead of silently ignoring error.